pkg/apis/core/types.go

```diff
@@ -3430,6 +3442,11 @@ type ServiceSpec struct {
 	// of peer discovery.
 	// +optional
 	PublishNotReadyAddresses bool
+
+	// The IPFamily value dictates the IPFamily of ClusterIP, LoadBalancerIP, Ingress IPs
```

Comment:
I had enough small comments that I'll just join them below.

First, is this an assertion that a Service supports JUST one family, or is it an expression of preference?  E.g. If I say IPv6, controllers will allocate v6 addresses, but is it required that `externalIPs` (which is user-provided and not controller-allocated) not have different families?  Will NodePorts only be forwarded from VM addresses of that particular family? 

```
ipFamily specifies whether this Service has a preference for a particular 
IP family (e.g. IPv4 vs. IPv6).  If a specific IP family is specified, addresses 
of that family will be required for the clusterIP, loadBalancerIP, 
loadBalancerSourceRanges, externalIPs, and ports[*].nodePort fields as 
well as for populating associated endpoints.  If no IP family is specified, 
the cluster-wide default will be used for the clusterIP, loadBalancerIP, and loadBalancerSourceRanges fields, the externalIPs field and associated 
endpoints may contain values from multiple IP families, and node ports 
will be bound to addresses regardless of IP family.  This field is immutable 
after creation. 
+optional
```

Check that for correctness of intention, but I think we need to spell out in PAINFUL detail what this means for the whole Service API.

Reply:
Also we should spell out what happens if I ask for v6 and the cluster does not have v6

Reply:
> Also we should spell out what happens if I ask for v6 and the cluster does not have v6

Same what happens today if the user assigned a clusterIP of different family (or out side the range) an error event will be broadcasted ref'ing service object and will ask the user to delete and recreate

Reply:
spell it out in the comments :)